<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

In [580]:
pip install psycopg2

Note: you may need to restart the kernel to use updated packages.


In [581]:
import pandas as pd
import psycopg2


In [583]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [586]:
# текст запроса
query_3_1 = '''
  SELECT COUNT(DISTINCT id)
    FROM vacancies
'''


In [587]:
# результат запроса
result_3_1=pd.read_sql_query(query_3_1, connection)
result_3_1

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\2279918825.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_1=pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [589]:
# текст запроса
query_3_2 = '''
  SELECT COUNT(DISTINCT id)
    FROM employers
'''

In [590]:
# результат запроса
result_3_2=pd.read_sql_query(query_3_2, connection)
result_3_2

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\210345736.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_2=pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [592]:
# текст запроса
query_3_3 = '''
  SELECT COUNT(DISTINCT id)
    FROM areas
'''

In [593]:
# результат запроса
result_3_3=pd.read_sql_query(query_3_3, connection)
result_3_3

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\3942791697.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_3=pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [595]:
# текст запроса
query_3_4 = '''
  SELECT COUNT(DISTINCT id)
    FROM industries
'''

In [596]:
# результат запроса
result_3_4=pd.read_sql_query(query_3_4, connection)
result_3_4

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\3114446276.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_4=pd.read_sql_query(query_3_4, connection)


count
0    294

***

По предварительному анализу выяснилось, что вакансии представлены более чем из тысячи регионов и почти 300 различных сфер. 
Как и ожидалось, количество вакансий превышает количество работодателей, и на каждую компанию в среднем приходится по две вакансии.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [601]:
# текст запроса
query_4_1 = '''
SELECT a.name area, COUNT(v.id) cnt
FROM vacancies v
JOIN areas a ON v.area_id=a.id
GROUP BY 1
ORDER BY COUNT(v.id) DESC
LIMIT 5
'''

In [602]:
result_4_1=pd.read_sql_query(query_4_1, connection)
result_4_1

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\3026521991.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_1=pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [604]:
# текст запроса
query_4_2 = '''
SELECT COUNT(*) 
FROM vacancies 
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [605]:
# результат запроса
result_4_2=pd.read_sql_query(query_4_2, connection)
result_4_2

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\345800217.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_2=pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [607]:
# текст запроса
query_4_3 = '''
SELECT  AVG(salary_from), AVG(salary_to)
FROM vacancies 
WHERE salary_to IS NOT NULL OR salary_from IS NOT NULL
'''

In [608]:
result_4_3=pd.read_sql_query(query_4_3, connection)
result_4_3

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\2185032766.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_3=pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [610]:
# текст запроса
query_4_4 = '''
SELECT schedule, employment, COUNT(*)
FROM vacancies
GROUP BY 1, 2
ORDER BY 3 DESC
'''

In [611]:
result_4_4=pd.read_sql_query(query_4_4, connection)
result_4_4

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\3109482894.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_4=pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [613]:
# текст запроса
query_4_5 = '''
SELECT experience, COUNT(*)
FROM vacancies
GROUP BY 1
ORDER BY 2 ASC
'''

In [614]:
result_4_5=pd.read_sql_query(query_4_5, connection)
result_4_5

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\943726770.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_5=pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Вакансии чаще всего размещаются из столиц стран СНГ или крупных городов. 
Почти в половине вакансий указан хотя бы один уровень зарплаты. 
Полный рабочий день и удаленная работа с полной занятостью пользуются большой популярностью. 
Наибольший спрос наблюдается на специалистов с опытом 1-3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [619]:
# текст запроса
query_5_1 = '''
SELECT e.name, COUNT(v.id)
FROM employers e
JOIN vacancies v ON e.id=v.employer_id
GROUP BY 1 
ORDER BY 2 DESC
LIMIT 5
'''

In [620]:
# результат запроса
result_5_1=pd.read_sql_query(query_5_1, connection)
result_5_1

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\2051332779.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_1=pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [622]:
# текст запроса
query_5_2 = '''
SELECT a.name, COUNT(e.id), COUNT(v.id)
FROM areas a 
LEFT JOIN employers e ON e.area=a.id
LEFT JOIN vacancies v ON v.area_id=a.id
WHERE v.id IS NULL 
GROUP BY 1
HAVING COUNT(e.id)>0
ORDER BY 2 DESC

'''

In [623]:
# результат запроса
result_5_2=pd.read_sql_query(query_5_2, connection)
result_5_2

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\3658117081.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_2=pd.read_sql_query(query_5_2, connection)


name  count  count
0                                       Россия    410      0
1                                    Казахстан    207      0
2                           Московская область     75      0
3                           Краснодарский край     19      0
4                                     Беларусь     18      0
..                                         ...    ...    ...
181                                  Миллерово      1      0
182  Михайловка (Уфимский район, Башкортостан)      1      0
183                                    Можайск      1      0
184                                Морозовск\n      1      0
185                                     Несвиж      1      0

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [625]:
# текст запроса
query_5_3 = '''
SELECT e.id, COUNT(DISTINCT v.area_id)
FROM employers e 
JOIN areas a ON a.id=e.area
JOIN vacancies v ON e.id=v.employer_id
GROUP BY e.id 
ORDER BY 2 DESC

'''

In [626]:
# результат запроса
result_5_3=pd.read_sql_query(query_5_3, connection)
result_5_3

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\2154887199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_3=pd.read_sql_query(query_5_3, connection)


id  count
0         1740    181
1         2748    152
2      5724811    116
3      5130287     88
4      3682876     71
...        ...    ...
14901   810278      1
14902   810313      1
14903   810551      1
14904   810688      1
14905   811902      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [628]:
# текст запроса
query_5_4 = '''
SELECT COUNT(DISTINCT e.id)
FROM employers e
LEFT JOIN employers_industries ei ON ei.employer_id=e.id
WHERE industry_id IS NULL

'''

In [629]:
# результат 
result_5_4=pd.read_sql_query(query_5_4, connection)
result_5_4

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\228748857.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_4=pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [631]:
# текст запроса
query_5_5 = '''
SELECT e.name, COUNT(ei.industry_id)
FROM employers e
JOIN employers_industries ei ON ei.employer_id=e.id
GROUP BY 1 
HAVING COUNT(ei.industry_id)=4
ORDER BY 1 ASC

'''

In [632]:
# результат запроса
result_5_5=pd.read_sql_query(query_5_5, connection)
result_5_5

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\3150796339.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_5=pd.read_sql_query(query_5_5, connection)


name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [634]:
# текст запроса
query_5_6 = '''
SELECT COUNT(DISTINCT e.id)
FROM employers e
JOIN employers_industries ei ON ei.employer_id=e.id
JOIN industries i ON i.id=ei.industry_id
WHERE i.name='Разработка программного обеспечения'
'''

In [635]:
# результат запроса
result_5_6=pd.read_sql_query(query_5_6, connection)
result_5_6

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\4172361763.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_6=pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [637]:
pip install requests beautifulsoup4


Note: you may need to restart the kernel to use updated packages.


In [638]:
# текст запроса
import requests
from bs4 import BeautifulSoup

url='https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response=requests.get(url)
page=BeautifulSoup(response.text,'html.parser')

# Extract city names
cities_here = page.find('table', class_="standard").find_all('a', title=True)
cities = [city['title'] for city in cities_here]

# Convert list of cities to SQL-friendly format
cities_sql = ', '.join([f"'{city}'" for city in cities])


In [639]:
query_5_7 = f'''
WITH city_vacancies AS (
    SELECT a.name, COUNT(v.id) AS cnt 
    FROM areas a
    JOIN vacancies v ON v.area_id = a.id
    JOIN employers e ON e.id = v.employer_id
    WHERE e.name = 'Яндекс'
      AND a.name IN ({cities_sql})
    GROUP BY a.name
    ORDER BY 2 ASC
)
SELECT * FROM city_vacancies

UNION ALL

SELECT 'Total', SUM(cnt)
FROM city_vacancies;
'''

# Execute the SQL query


In [640]:
# результат запроса
result_5_7 = pd.read_sql_query(query_5_7, connection)
result_5_7

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\786645949.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_7 = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

Россия — это регион с наибольшим количеством работодателей, однако с минимальным числом вакансий.
Яндекс — самая географически распространенная компания, с вакансиями в 181 регионе. 
8419 работодателей, что составляет 36%, не указывают сферу деятельности. 
3553 работодателя, или 15%, указывают сферу деятельности «Разработка программного обеспечения».

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [645]:
# текст запроса
query_6_1 = '''
SELECT COUNT(*)
FROM vacancies v 
WHERE v.name ILIKE '%data%' OR v.name ILIKE '%данн%'
'''

In [646]:
# результат запроса
result_6_1=pd.read_sql_query(query_6_1, connection)
result_6_1

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\1424470524.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_1=pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [648]:
# текст запроса
query_6_2 = '''
SELECT count(DISTINCT v.id) AS cnt
FROM vacancies v
WHERE (v.name ILIKE '%data scientist%'
OR v.name ILIKE '%data science%'
OR v.name ILIKE '%исследователь данных%'
OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
OR v.name ILIKE 'machine learning'
OR v.name ILIKE '%машинн%обучен%')
AND (v.name ILIKE '%junior%'
OR v.experience ILIKE '%Нет опыта%'
OR v.employment ILIKE '%Стажировка%')
'''

In [649]:
# результат запроса
result_6_2=pd.read_sql_query(query_6_2, connection)
result_6_2

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\2011958607.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_2=pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [651]:
# текст запроса
query_6_3 = '''
SELECT count(DISTINCT v.id) AS cnt
FROM vacancies v
WHERE (v.name ILIKE '%data scientist%'
OR v.name ILIKE '%data science%'
OR v.name ILIKE '%исследователь данных%'
OR v.name ILIKE '%machine learning%'
OR v.name ILIKE '%машинн%обучен%'
OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%'))
AND (v.key_skills ILIKE '%sql%'
OR v.key_skills ILIKE '%postgres%')
'''

In [652]:
# результат запроса
result_6_3=pd.read_sql_query(query_6_3, connection)
result_6_3

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\1621525378.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_3=pd.read_sql_query(query_6_3, connection)


cnt
0  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [654]:
# текст запроса
query_6_4 = '''
SELECT count(DISTINCT v.id) AS cnt
FROM vacancies v
WHERE (v.name ILIKE '%data scientist%'
OR v.name ILIKE '%data science%'
OR v.name ILIKE '%исследователь данных%'
OR v.name ILIKE '%machine learning%'
OR v.name ILIKE '%машинн%обучен%'
OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%'))
AND (v.key_skills ILIKE '%python%')
'''

In [655]:
# результат запроса
result_6_4=pd.read_sql_query(query_6_4, connection)
result_6_4

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\1180772487.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_4=pd.read_sql_query(query_6_4, connection)


cnt
0  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [657]:
# текст запроса
query_6_5 = '''
SELECT AVG(skill_count) AS avg_skills
FROM (
    SELECT LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, CHR(9), '')) + 1 AS skill_count
    FROM vacancies v
    WHERE (
        v.name ILIKE '%data scientist%'
        OR v.name ILIKE '%data science%'
        OR v.name ILIKE '%исследователь данных%'
        OR v.name ILIKE '%machine learning%'
        OR v.name ILIKE '%машинн%обучен%'
        OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%')
    )

) subquery;
'''

In [658]:
# результат запроса
result_6_5=pd.read_sql_query(query_6_5, connection)
result_6_5

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\1042119364.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_5=pd.read_sql_query(query_6_5, connection)


avg_skills
0    6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [660]:
# текст запроса
query_6_6 = '''
SELECT AVG(
           CASE
               WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
               WHEN salary_from IS NOT NULL THEN salary_from
               WHEN salary_to IS NOT NULL THEN salary_to
           END
       ) AS avg_salary
FROM vacancies v
WHERE (
    (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%'
    OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%'))
    AND  (v.experience='От 3 до 6 лет')
    
)
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL);

'''

In [661]:
# результат запроса
result_6_6=pd.read_sql_query(query_6_6, connection)
result_6_6

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\648168112.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_6=pd.read_sql_query(query_6_6, connection)


avg_salary
0  243114.666667

***

Из всех вакансий 1771, что составляет 3,6%, относятся к сфере данных. 
Среди них 51 вакансия (2,9%) предназначена для специалистов уровня junior. SQL требуется в 12,3% вакансий, связанных с данными, а Python — в 20,2%. 
Анализ показал, что специалисты по данным с опытом от 3 до 6 лет зарабатывают на 37% больше, 
чем в среднем среди всех специалистов в сфере данных (243 тыс. рублей против 177 тыс. рублей). 
По сравнению с общими средними зарплатами по всем вакансиям их зарплаты почти в три раза выше (243 тыс. рублей против 84 тыс. рублей).

# Общий вывод по проекту
Вакансии представлены в более чем тысяче регионов и охватывают почти 300 различных сфер, причем на каждую компанию в среднем приходится по две вакансии. Вакансии чаще всего размещаются в столицах стран СНГ или крупных городах, а почти в половине из них указаны зарплатные предложения. Полный рабочий день и удалённая работа с полной занятостью остаются наиболее востребованными форматами.

Наибольший спрос наблюдается на специалистов с опытом работы от 1 до 3 лет. Россия лидирует по количеству работодателей, однако предлагает минимальное количество вакансий. Яндекс является самой географически разнообразной компанией, с вакансиями в 181 регионе. При этом значительная часть работодателей (36%) не указывает сферу деятельности, а 15% относятся к разработке программного обеспечения.

Сфера данных составляет 3,6% от всех вакансий, из которых 2,9% ориентированы на специалистов уровня junior. SQL и Python остаются важными навыками: SQL упоминается в 12,3% вакансий по данным, а Python — в 20,2%. Специалисты по данным с опытом от 3 до 6 лет зарабатывают на 37% больше, чем в среднем по рынку данных (243 тыс. рублей против 177 тыс. рублей), что почти в три раза выше средней зарплаты по всем вакансиям (243 тыс. рублей против 84 тыс. рублей).

In [665]:
# Средняя зарплата базы вакансий
query_7_1 = '''
SELECT AVG(
           CASE
               WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
               WHEN salary_from IS NOT NULL THEN salary_from
               WHEN salary_to IS NOT NULL THEN salary_to
           END
       ) AS avg_salary
FROM vacancies v
WHERE  (salary_from IS NOT NULL OR salary_to IS NOT NULL);
'''


In [666]:
result_7_1=pd.read_sql_query(query_7_1, connection)
result_7_1

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\1541841109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_7_1=pd.read_sql_query(query_7_1, connection)


avg_salary
0  84314.865326

In [667]:
# Средняя зарплата специалистов по данным
query_7_2 = '''
SELECT AVG(
           CASE
               WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
               WHEN salary_from IS NOT NULL THEN salary_from
               WHEN salary_to IS NOT NULL THEN salary_to
           END
       ) AS avg_salary
FROM vacancies v
WHERE (
    (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%'
    OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%'))
    
    
)
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL);
'''


In [668]:
result_7_2=pd.read_sql_query(query_7_2, connection)
result_7_2

C:\Users\UserPC\AppData\Local\Temp\ipykernel_14936\1772489806.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_7_2=pd.read_sql_query(query_7_2, connection)


avg_salary
0  177378.854839